feat: support for noEmitOnErrors property in webpack.config

Fatme opened the pull request:
Currently [noEmitOnErrors property](https://webpack.js.org/configuration/optimization/#optimizationnoemitonerrors) is not respected from webpack compilation in case it is set in `webpack.config.js`. As a result, the files are emitted even when there is an error in the source code. However, it turns out that `ts-loader` is working correctly in `transpileOnly: false` mode and the problem is only in case when `ts-loader` is started in `transpileOnly: true` + `ForkTsCheckerWebpackPlugin`.

After investigating the issue, it turns out that the problem is [here](https://github.com/TypeStrong/fork-ts-checker-webpack-plugin/blob/4e8c6f98fb14dced5f40a56723b459e519d59362/src/index.ts#L574). Webpack has a logic that checks if the files should be emitted based on `compilation.errors` array. This logic is executed on `shouldEmit` hook as can be seen [here](https://github.com/webpack/webpack/blob/95d21bb39a1e700f28be776b300492c50a169fc7/lib/NoEmitOnErrorsPlugin.js#L9-L11). So, we need to put our reported `diagnostics` and `lints` in `compilation.errors` **before** `shouldEmit`. 

The `afterCompile` hook seems as a perfect solution as it is executed before `shouldEmit` and is an `async` hook. As supporting this feature is important for our error handling story, we implemented the change, spent a little time testing it and it works for our cases.

However, we're not absolutely sure for the impact of the mentioned change so we'd be glad to hear your opinion and suggestions.

Rel to: https://github.com/TypeStrong/fork-ts-checker-webpack-plugin/issues/57

Fatme committed:
fix: replace emit hook with afterCompile

The afterCompile hook should be blocked instead of emit in order to handle correctly noEmitOnErrors property from webpack.config.js

BREAKING CHANGE: The emit hook is replaced with afterCompile hook.

Fatme committed:
refactor: rename emitCallback to afterCompileCallback

piotr-oles commented:
Thank you for your findings and for the PR :)
Unfortunately, changing the hook from `emit` to `afterCompile` is a breaking change in my opinion. There is a custom hook: `fork-ts-checker-emit` - and as it's a part of public API, some developers could assume that it will run on the `emit` event.
Because of that, we need to find a method to deprecate the `fork-ts-checker-emit` hook and add a new one (`fork-ts-checker-after-compile`) + keep default behavior of using webpack's `emit` event (probably we will have to add a new configuration option to control that behavior).

I know that it can be frustrating that we cannot just fix this bug and change the hook, but our responsibility is to introduce breaking changes gracefully :)

Fatme commented:
Hey @piotr-oles,

Thank you very much for taking a look at this PR.

> but our responsibility is to introduce breaking changes gracefully

I totally agree that we have to ensure seamless transition for the users.

> keep default behavior of using webpack's emit event (probably we will have to add a new configuration option to control that behavior)

Yep, maybe introducing a new configuration option is a good solution. Having that option with default value false will ensure the new version of the plugin will not break existing users and applications. So we need a name for the new option 😸 - maybe `useAfterCompile` or `noEmitOnErrors` or something else? I'm opened to any ideas and suggestions.

As this PR is really important for us and our error handling story, I'll prepare the needed changes once we have an agreement about the name of the new option.

piotr-oles commented:
I would suggest `useAfterCompileHook`, @johnnyreilly, what do you think?

johnnyreilly commented:
We could do... I'm just wondering if it might make more sense to do a breaking changes version? After all, do we want to maintain two behaviours in the long term? If we all agree this is a more sensible behaviour then having a breaking changes version seems like a way to not break existing users and move the plugin in the direction we want to go... Thoughts?

Fatme commented:
Hey @piotr-oles, @johnnyreilly

It seems we have 2 options:
* Release a breaking change version of the plugin - If we agree on that, I suppose it'll be released as a major version e.g `2.0.0`.
* Add a configuration option and support the both behaviors.

The both solutions will work for us and will address our problem. However, personally to me, I prefer releasing a major version and introducing it as a breaking change. That will be the most beneficial solution in the long term.

Let me know about your decision, so I can apply the required changes on the PR.

piotr-oles commented:
Ok, maybe breaking change is a better option - it will probably not break anything for 99% of users.

Please update the documentation (Plugin hooks section) and add [BREAKING CHANGE](https://github.com/TypeStrong/fork-ts-checker-webpack-plugin/blob/master/CONTRIBUTING.md#examples) section to the commit message :)

Fatme committed:
fix: change fork-ts-checker-emit to fork-ts-checker-after-compile

Fatme commented:
@piotr-oles,

I changed the commit message with BREAKING CHANGE: footer and updated the plugin section in the documentation.

Let me know if I've missed something or there is something else that I can do.

johnnyreilly commented:
Could you catch up this branch with master please?

Fatme commented:
@piotr-oles, @johnnyreilly,

I've rebased the PR on the latest master branch.

johnnyreilly reviewed (dismissed):
Looks good to me!

Fatme commented:
@johnnyreilly, @piotr-oles,

It seems there are failing CI builds https://travis-ci.org/TypeStrong/fork-ts-checker-webpack-plugin/jobs/602424646?utm_medium=notification&utm_source=github_status:
```
error An unexpected error occurred: "https://registry.yarnpkg.com/babel-code-frame/-/babel-code-frame-6.26.0.tgz: Request failed \"503 Service Unavailable\"".
info If you think this is a bug, please open a bug report with the information provided in "/home/travis/build/TypeStrong/fork-ts-checker-webpack-plugin/yarn-error.log".
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
The command "yarn install --frozen-lockfile" failed and exited with 1 during .
```

Maybe we should rerun the travis builds.

johnnyreilly commented:
Retriggered. Travis has been pretty flaky this week

Fatme commented:
@johnnyreilly, @piotr-oles 

We have a green build! 🎉 

Do I need to rebase it again on master as there are additional commits merged into master or you can merge the PR directly?

johnnyreilly commented:
Yes please - update the branch!

Fatme commented:
@piotr-oles, @johnnyreilly,

Thank you very much for merging the PR!
When do you plan to release it officially?

piotr-oles commented:
It will be released automatically by [`📦 🚀 semantic-release`](https://github.com/semantic-release/semantic-release) when CI build finish :)

Fatme commented:
@piotr-oles,

What about the version - will it be updated to `2.0.0` as we're releasing a breaking change?

piotr-oles commented:
:tada: This PR is included in version 2.0.0 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/fork-ts-checker-webpack-plugin)
- [GitHub release](https://github.com/TypeStrong/fork-ts-checker-webpack-plugin/releases/tag/v2.0.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:

piotr-oles commented:
The CI is performing very badly last time - I had to re-run the release job. And yes - as you can see semantic-release do all the magic ✨ 😃

Fatme commented:
Yeep, it's really cool magic! 😃
